src/vmspawn/vmspawn-scope.c

```diff
@@ -0,0 +1,426 @@
+}
+
+/* NOTE: this function should really take cmdline as `const char * const *` but cannot until sd_bus_message_append_str does as well */
+int attach_command_to_socket_in_scope(sd_bus *bus, const char *scope, const char *unit_name, const char *socket_path,
+                                      int socket_type, char **cmdline, char **cleanup_cmdline, char **extra_properties) {
+        _cleanup_(bus_wait_for_jobs_freep) BusWaitForJobs *w = NULL;
+        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+        _cleanup_(sd_bus_message_unrefp) sd_bus_message *m = NULL, *reply = NULL;
+        _cleanup_free_ char *service_desc = NULL, *service_name = NULL, *socket_name = NULL, *socket_desc = NULL;
+        const char *object, *socket_type_str;
+        int r;
+
+        /*
+         * Starts a socket/service unit pair bound to the given scope.
+         *
+         * The socket unit listens on `socket_path` with the for connections of type `socket_type`.
+         * When a connection to the socket is initiated the service is started to and passed the socket's file descriptor as fd 3.
+         *
+         * `cmdline` is forwarded to the service unit's ExecStart property.
+         * `cleanup_cmdline` is forwarded to the service unit's ExecStopPost property.
+         *
+         * `extra_properties` is used to set any extra properties this service unit should have.
+         * The format is the same as in unit files, e.g. Requires=other-unit.socket
+         */
+
+        assert(bus);
+        assert(scope);
+        assert(unit_name);
+        assert(socket_path);
+        assert(cmdline);
+        assert_return(IN_SET(socket_type, SOCK_STREAM, SOCK_SEQPACKET, SOCK_DGRAM), -EOPNOTSUPP);
+
+        r = bus_wait_for_jobs_new(bus, &w);
+        if (r < 0)
+                return log_error_errno(r, "Could not watch job: %m");
+
+        r = unit_name_mangle_with_suffix(unit_name, "as socket", 0, ".socket", &socket_name);
+        if (r < 0)
+                return log_error_errno(r, "Failed to mangle unit name %s: %m", unit_name);
+
+        r = unit_name_change_suffix(socket_name, ".service", &service_name);
+        if (r < 0)
+                return log_error_errno(r, "Failed to change unit suffix %s: %m", unit_name);
+
+        service_desc = strv_join(cmdline, " ");
+        if (!service_desc)
+                return log_oom();
+
+        socket_desc = strjoin("socket: ", socket_path);
+        if (!socket_desc)
+                return log_oom();
+
+        switch (socket_type) {
+        case SOCK_STREAM:
+                socket_type_str = "Stream";
+                break;
+        case SOCK_DGRAM:
+                socket_type_str = "Datagram";
+                break;
+        case SOCK_SEQPACKET:
+                socket_type_str = "SequentialPacket";
+                break;
+        default:
+                return log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "Invalid socket type: %d", socket_type);
+        }
```

Comment:
can you move the translation of the socket type to some helper please? i.e. one of those table-based xyz_to_string( )helpers

Reply:
I found a lookup table for SOCK_* entries in socket-util.c so I'm going to move to using that as I think there is no need for a function once its a table lookup.

Reply:
Fixed in eae47ee9d9.